File: examples/typescript/missionControl/agent.ts
#!/usr/bin/env bun
/**
 * Example agent that connects to Mission Control.
 *
 * Demonstrates proxy client usage (Chapters 1-3):
 *   - getStatus (unary)
 *   - submitLog (unary)
 *   - ingestMetrics (client streaming)
 *
 * Usage:
 *     bun run agent.ts <server-address>
 */

import { AsterClientWrapper } from '@aster-rpc/aster';

const address = process.argv[2];
if (!address) {
  console.error("Usage: bun run agent.ts <server-address>");
  process.exit(1);
}

const client = new AsterClientWrapper({ address });
await client.connect();
const mc = client.proxy("MissionControl");

// Chapter 1: check in
const status = await mc.getStatus({ agentId: "ts-agent-1" });
console.log("Status:", status);

// Chapter 2: push a log entry
await mc.submitLog({
  timestamp: Date.now() / 1000,
  level: "info",
  message: "agent started",
  agentId: "ts-agent-1",
});
console.log("Log submitted");

// Chapter 3: stream 1000 metrics
async function* metrics() {
  for (let i = 0; i < 1000; i++) {
    yield {
      name: "cpu.usage",
      value: Math.random() * 100,
      timestamp: Date.now() / 1000,
    };
  }
}

const result = await mc.ingestMetrics(metrics());
console.log(`Metrics accepted: ${(result as any).accepted}`);

await client.close();

File: examples/python/mission_control/agent.py
#!/usr/bin/env python3
"""Example agent that connects to Mission Control.

Demonstrates proxy client usage (Chapters 1-3):
  - getStatus (unary)
  - submitLog (unary)
  - ingestMetrics (client streaming)

Usage:
    python -m mission_control.agent <server-address>
"""

import asyncio
import sys
import time
from random import random

from aster import AsterClient


async def run(address: str) -> None:
    client = AsterClient(address=address)
    await client.connect()
    mc = client.proxy("MissionControl")

    # Chapter 1: check in
    status = await mc.getStatus({"agent_id": "py-agent-1"})
    print(f"Status: {status}")

    # Chapter 2: push a log entry
    await mc.submitLog({
        "timestamp": time.time(),
        "level": "info",
        "message": "agent started",
        "agent_id": "py-agent-1",
    })
    print("Log submitted")

    # Chapter 3: stream 1000 metrics
    async def metrics():
        for _ in range(1000):
            yield {
                "name": "cpu.usage",
                "value": random() * 100,
                "timestamp": time.time(),
            }

    result = await mc.ingestMetrics(metrics())
    print(f"Metrics accepted: {result['accepted']}")

    await client.close()


def main() -> None:
    if len(sys.argv) < 2:
        print("Usage: python -m mission_control.agent <server-address>")
        sys.exit(1)
    asyncio.run(run(sys.argv[1]))


if __name__ == "__main__":
    main()
